eth_p2p/rlpx_protocols/shh_protocol.nim

```diff
@@ -416,31 +539,406 @@ proc add*(self: var Queue, msg: Message) =
       if last.pow > msg.pow or
         (last.pow == msg.pow and last.env.expiry > msg.env.expiry):
         # The new message has less pow or will expire earlier - drop it
-        self.items.del(self.items.len() - 1)
+        return false
 
-  self.items.insert(msg, self.items.lowerBound(msg, cmpPow))
+      self.items.del(self.items.len() - 1)
+      self.itemHashes.excl(last)
+
+  # check for duplicate
+  # NOTE: Could also track if duplicates come from the same peer and disconnect
+  # from that peer. Is this tracking overhead worth it though?
+  if self.itemHashes.containsOrIncl(msg):
+    return false
+  else:
+    self.items.insert(msg, self.items.lowerBound(msg, cmpPow))
+    return true
+
+# Filters ----------------------------------------------------------------------
+proc newFilter*(src = none[PublicKey](), privateKey = none[PrivateKey](),
+                symKey = none[SymKey](), topics: seq[Topic] = @[],
+                powReq = 0.0, allowP2P = false): Filter =
+  Filter(src: src, privateKey: privateKey, symKey: symKey, topics: topics,
+         powReq: powReq, allowP2P: allowP2P, bloom: toBloom(topics))
+
+proc subscribeFilter*(filters: var Filters, filter: Filter,
+                      handler = none[FilterMsgHandler]()): string =
```

Comment:
You could have gotten away by treating the `nil` pointer value as the equivalent of `none`.

Reply:
You mean set `nil` as default and check for this in the code?